src/fabric/src/fabric2_db.erl

```diff
@@ -969,6 +970,145 @@ fold_docs(Db, UserFun, UserAcc0, Options) ->
     end).
 
 
+fold_docs(Db, DocIds, UserFun, UserAcc0, Options) ->
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        try
+            NeedsTreeOpts = [revs_info, conflicts, deleted_conflicts],
+            NeedsTree = (Options -- NeedsTreeOpts /= Options),
+
+            FetchRevs = case NeedsTree of
+                true ->
+                    fun(DocId) ->
+                        fabric2_fdb:get_all_revs_future(TxDb, DocId)
+                    end;
+                false ->
+                    fun(DocId) ->
+                        fabric2_fdb:get_winning_revs_future(TxDb, DocId, 1)
+                    end
+            end,
+            InitAcc = #{
+                revs_q => queue:new(),
+                revs_count => 0,
+                body_q => queue:new(),
+                body_count => 0,
+                doc_opts => Options,
+                user_acc => UserAcc0,
+                user_fun => UserFun
+            },
+
+            FinalAcc1 = lists:foldl(fun(DocId, Acc) ->
+                #{
+                    revs_q := RevsQ,
+                    revs_count := RevsCount
+                } = Acc,
+                Future = FetchRevs(DocId),
+                NewAcc = Acc#{
+                    revs_q := queue:in({DocId, Future}, RevsQ),
+                    revs_count := RevsCount + 1
+                },
+                drain_fold_docs_revs_futures(TxDb, NewAcc)
+            end, InitAcc, DocIds),
+
+            FinalAcc2 = drain_all_fold_docs_revs_futures(TxDb, FinalAcc1),
+            FinalAcc3 = drain_all_fold_docs_body_futures(TxDb, FinalAcc2),
+
+            #{
+                user_acc := FinalUserAcc
+            } = FinalAcc3,
+            {ok, FinalUserAcc}
+
+        catch throw:{stop, StopUserAcc} ->
+            {ok, StopUserAcc}
+        end
+    end).
+
+
+drain_fold_docs_revs_futures(_TxDb, #{revs_count := C} = Acc) when C < 100 ->
```

Comment:
All of these support functions should be moved to the end of the module. I'd put them just after `drain_all_deleted_info_futures`.

Reply:
Fixed